File: jdbc/src/main/java/org/enodeframework/jdbc/JDBCEventStore.java
package org.enodeframework.jdbc;

import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.jdbc.JDBCClient;
import io.vertx.ext.sql.SQLClient;
import io.vertx.ext.sql.SQLConnection;
import org.enodeframework.common.exception.EventStoreException;
import org.enodeframework.common.exception.IORuntimeException;
import org.enodeframework.common.io.IOHelper;
import org.enodeframework.common.serializing.ISerializeService;
import org.enodeframework.common.utilities.Ensure;
import org.enodeframework.eventing.AggregateEventAppendResult;
import org.enodeframework.eventing.BatchAggregateEventAppendResult;
import org.enodeframework.eventing.DomainEventStream;
import org.enodeframework.eventing.EventAppendResult;
import org.enodeframework.eventing.EventAppendStatus;
import org.enodeframework.eventing.IEventSerializer;
import org.enodeframework.eventing.IEventStore;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.sql.DataSource;
import java.sql.SQLException;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

/**
 * @author anruence@gmail.com
 */
public abstract class JDBCEventStore extends AbstractVerticle implements IEventStore {

    private static final Logger logger = LoggerFactory.getLogger(JDBCEventStore.class);

    private static final String EVENT_TABLE_NAME_FORMAT = "%s_%s";

    private static final String INSERT_EVENT_SQL = "INSERT INTO %s (aggregate_root_id, aggregate_root_type_name, command_id, version, gmt_create, events) VALUES (?, ?, ?, ?, ?, ?)";
    private static final String SELECT_MANY_BY_VERSION_SQL = "SELECT * FROM %s WHERE aggregate_root_id = ? AND version >= ? AND Version <= ? ORDER BY version";
    private static final String SELECT_ONE_BY_VERSION_SQL = "SELECT * FROM %s WHERE aggregate_root_id = ? AND version = ?";
    private static final String SELECT_ONE_BY_COMMAND_ID_SQL = "SELECT * FROM %s WHERE aggregate_root_id = ? AND command_id = ?";
    private static final String SELECT_MAX_VERSION_SQL = "SELECT max(version) AS version FROM %s WHERE aggregate_root_id = ? AND aggregate_root_type_name = ?";

    private final String tableName;
    private final int tableCount;
    private final String sqlState;
    private final String versionIndexName;
    private final String commandIndexName;
    /**
     * dataSource 如果使用了分库分表的ShardDataSource，分表则不再需要
     */
    private final DataSource dataSource;
    private final IEventSerializer eventSerializer;
    private final ISerializeService serializeService;
    private SQLClient sqlClient;

    public JDBCEventStore(DataSource dataSource, IEventSerializer eventSerializer, ISerializeService serializeService) {
        this(dataSource, new DBConfiguration(), eventSerializer, serializeService);
    }

    public JDBCEventStore(DataSource dataSource, DBConfiguration dbConfiguration, IEventSerializer eventSerializer, ISerializeService serializeService) {
        this.serializeService = serializeService;
        Ensure.notNull(dataSource, "dataSource");
        Ensure.notNull(eventSerializer, "eventSerializer");
        Ensure.notNull(dbConfiguration, "dbConfiguration");
        this.eventSerializer = eventSerializer;
        this.dataSource = dataSource;
        this.tableName = dbConfiguration.getEventTableName();
        this.tableCount = dbConfiguration.getEventTableCount();
        this.sqlState = dbConfiguration.getSqlState();
        this.versionIndexName = dbConfiguration.getEventTableVersionUniqueIndexName();
        this.commandIndexName = dbConfiguration.getEventTableCommandIdUniqueIndexName();
    }

    @Override
    public void start() {
        sqlClient = JDBCClient.create(vertx, dataSource);
    }

    @Override
    public CompletableFuture<EventAppendResult> batchAppendAsync(List<DomainEventStream> eventStreams) {
        CompletableFuture<EventAppendResult> future = new CompletableFuture<>();
        EventAppendResult appendResult = new EventAppendResult();
        if (eventStreams.size() == 0) {
            future.complete(appendResult);
            return future;
        }
        Map<String, List<DomainEventStream>> eventStreamMap = eventStreams.stream().distinct().collect(Collectors.groupingBy(DomainEventStream::getAggregateRootId));
        BatchAggregateEventAppendResult batchAggregateEventAppendResult = new BatchAggregateEventAppendResult(eventStreamMap.keySet().size());
        for (Map.Entry<String, List<DomainEventStream>> entry : eventStreamMap.entrySet()) {
            batchAppendAggregateEventsAsync(entry.getKey(), entry.getValue(), batchAggregateEventAppendResult, 0);
        }
        return batchAggregateEventAppendResult.taskCompletionSource;
    }

    private void batchAppendAggregateEventsAsync(String aggregateRootId, List<DomainEventStream> eventStreamList, BatchAggregateEventAppendResult batchAggregateEventAppendResult, int retryTimes) {
        IOHelper.tryAsyncActionRecursively("BatchAppendAggregateEventsAsync",
                () -> batchAppendAggregateEventsAsync(aggregateRootId, eventStreamList),
                result -> {
                    batchAggregateEventAppendResult.addCompleteAggregate(aggregateRootId, result);
                },
                () -> String.format("[aggregateRootId: %s, eventStreamCount: %s]", aggregateRootId, eventStreamList.size()),
                null,
                retryTimes, true);
    }

    private CompletableFuture<DomainEventStream> tryFindEventByCommandIdAsync(String aggregateRootId, String commandId, List<String> duplicateCommandIds, int retryTimes) {
        CompletableFuture<DomainEventStream> future = new CompletableFuture<>();
        IOHelper.tryAsyncActionRecursively("TryFindEventByCommandIdAsync",
                () -> findAsync(aggregateRootId, commandId),
                result -> {
                    if (result != null) {
                        duplicateCommandIds.add(result.getCommandId());
                    }
                    future.complete(result);
                },
                () -> String.format("[aggregateRootId:%s, commandId:%s]", aggregateRootId, commandId),
                null,
                retryTimes, true);
        return future;
    }

    private CompletableFuture<AggregateEventAppendResult> batchAppendAggregateEventsAsync(String aggregateRootId, List<DomainEventStream> eventStreamList) {
        String sql = String.format(INSERT_EVENT_SQL, getTableName(aggregateRootId));
        List<JsonArray> jsonArrays = Lists.newArrayList();
        eventStreamList.sort(Comparator.comparingInt(DomainEventStream::getVersion));
        for (DomainEventStream domainEventStream : eventStreamList) {
            JsonArray array = new JsonArray();
            array.add(domainEventStream.getAggregateRootId());
            array.add(domainEventStream.getAggregateRootTypeName());
            array.add(domainEventStream.getCommandId());
            array.add(domainEventStream.getVersion());
            array.add(domainEventStream.getTimestamp().toInstant());
            array.add(serializeService.serialize(eventSerializer.serialize(domainEventStream.events())));
            jsonArrays.add(array);
        }
        CompletableFuture<AggregateEventAppendResult> future = batchInsertAsync(sql, jsonArrays);
        return future.exceptionally(throwable -> {
            if (throwable instanceof SQLException) {
                SQLException ex = (SQLException) throwable;
                if (ex.getSQLState().equals(sqlState) && ex.getMessage().contains(versionIndexName)) {
                    AggregateEventAppendResult appendResult = new AggregateEventAppendResult();
                    appendResult.setEventAppendStatus(EventAppendStatus.DuplicateEvent);
                    return appendResult;
                }
                if (ex.getSQLState().equals(sqlState) && ex.getMessage().contains(commandIndexName)) {
                    // Duplicate entry '5d3ac841d1fcfe669e9a257d-5d3ac841d1fcfe669e9a2585' for key 'IX_EventStream_AggId_CommandId'
                    // ERROR: duplicate key value violates unique constraint "event_stream_aggregate_root_id_command_id_key"\n详细：Key (aggregate_root_id, command_id)=(5ee99656d767113d73a7540f, 5ee99656d767113d73a75417) already exists.
                    AggregateEventAppendResult appendResult = new AggregateEventAppendResult();
                    appendResult.setEventAppendStatus(EventAppendStatus.DuplicateCommand);
                    String commandId = parseDuplicateCommandId(ex.getMessage());
                    if (!Strings.isNullOrEmpty(commandId)) {
                        appendResult.setDuplicateCommandIds(Lists.newArrayList(commandId));
                        return appendResult;
                    }
                    // 如果没有从异常信息获取到commandId(很低概率获取不到)，兜底从db查询出来，此时选择的策略是从最后一个开始查
                    // 但是Vert.x的线程模型决定了不能再次使用eventloop线程执行阻塞的查询操作
                    tryFindEventByCommandIdAsyncRecursion(eventStreamList.size(), aggregateRootId, eventStreamList, appendResult.getDuplicateCommandIds(), 0);
                    return appendResult;
                }
                logger.error("Batch append event has sql exception.", throwable);
                throw new IORuntimeException(throwable);
            }
            logger.error("Batch append event has unknown exception.", throwable);
            throw new EventStoreException(throwable);
        });
    }

    private void tryFindEventByCommandIdAsyncRecursion(int i, String aggregateRootId, List<DomainEventStream> eventStreamList, List<String> duplicateCommandIds, int retryTimes) {
        if (i == 0) {
            return;
        }
        DomainEventStream eventStream = eventStreamList.get(i - 1);
        tryFindEventByCommandIdAsync(aggregateRootId, eventStream.getCommandId(), duplicateCommandIds, retryTimes).thenAccept(x -> {
            if (x != null) {
                return;
            }
            tryFindEventByCommandIdAsyncRecursion(i - 1, aggregateRootId, eventStreamList, duplicateCommandIds, 0);
        });
    }

    abstract protected String parseDuplicateCommandId(String msg);

    public CompletableFuture<AggregateEventAppendResult> batchInsertAsync(String sql, List<JsonArray> jsonArrays) {
        CompletableFuture<AggregateEventAppendResult> future = new CompletableFuture<>();
        batchWithParams(sql, jsonArrays, x -> {
            if (x.succeeded()) {
                AggregateEventAppendResult appendResult = new AggregateEventAppendResult();
                appendResult.setEventAppendStatus(EventAppendStatus.Success);
                future.complete(appendResult);
                return;
            }
            future.completeExceptionally(x.cause());
        });
        return future;
    }

    @Override
    public CompletableFuture<List<DomainEventStream>> queryAggregateEventsAsync(String aggregateRootId, String aggregateRootTypeName, int minVersion, int maxVersion) {
        return IOHelper.tryIOFuncAsync(() -> {
            CompletableFuture<List<DomainEventStream>> future = new CompletableFuture<>();
            String sql = String.format(SELECT_MANY_BY_VERSION_SQL, getTableName(aggregateRootId));
            JsonArray array = new JsonArray();
            array.add(aggregateRootId);
            array.add(minVersion);
            array.add(maxVersion);
            sqlClient.queryWithParams(sql, array, x -> {
                if (x.succeeded()) {
                    List<StreamRecord> results = Lists.newArrayList();
                    x.result().getRows().forEach(row -> results.add(row.mapTo(StreamRecord.class)));
                    List<DomainEventStream> streams = results.stream().map(this::convertFrom).collect(Collectors.toList());
                    future.complete(streams);
                    return;
                }
                future.completeExceptionally(x.cause());
            });
            return future.exceptionally(throwable -> {
                if (throwable instanceof SQLException) {
                    SQLException ex = (SQLException) throwable;
                    String errorMessage = String.format("Failed to query aggregate events async, aggregateRootId: %s, aggregateRootType: %s", aggregateRootId, aggregateRootTypeName);
                    logger.error(errorMessage, ex);
                    throw new IORuntimeException(throwable);
                }
                logger.error("Failed to query aggregate events async, aggregateRootId: {}, aggregateRootType: {}", aggregateRootId, aggregateRootTypeName, throwable);
                throw new EventStoreException(throwable);
            });
        }, "QueryAggregateEventsAsync");
    }

    @Override
    public CompletableFuture<DomainEventStream> findAsync(String aggregateRootId, int version) {
        return IOHelper.tryIOFuncAsync(() -> {
            CompletableFuture<DomainEventStream> future = new CompletableFuture<>();
            String sql = String.format(SELECT_ONE_BY_VERSION_SQL, getTableName(aggregateRootId));
            JsonArray array = new JsonArray();
            array.add(aggregateRootId);
            array.add(version);
            sqlClient.queryWithParams(sql, array, x -> {
                if (x.succeeded()) {
                    DomainEventStream stream = null;
                    Optional<JsonObject> first = x.result().getRows().stream().findFirst();
                    if (first.isPresent()) {
                        StreamRecord record = first.get().mapTo(StreamRecord.class);
                        stream = convertFrom(record);
                    }
                    future.complete(stream);
                    return;
                }
                future.completeExceptionally(x.cause());
            });
            return future.exceptionally(throwable -> {
                if (throwable instanceof SQLException) {
                    SQLException ex = (SQLException) throwable;
                    logger.error("Find event by version has sql exception, aggregateRootId: {}, version: {}", aggregateRootId, version, ex);
                    throw new IORuntimeException(throwable);
                }
                logger.error("Find event by version has unknown exception, aggregateRootId: {}, version: {}", aggregateRootId, version, throwable);
                throw new EventStoreException(throwable);
            });
        }, "FindEventByVersionAsync");

    }

    @Override
    public CompletableFuture<DomainEventStream> findAsync(String aggregateRootId, String commandId) {
        return IOHelper.tryIOFuncAsync(() -> {
            CompletableFuture<DomainEventStream> future = new CompletableFuture<>();
            String sql = String.format(SELECT_ONE_BY_COMMAND_ID_SQL, getTableName(aggregateRootId));
            JsonArray array = new JsonArray();
            array.add(aggregateRootId);
            array.add(commandId);
            sqlClient.queryWithParams(sql, array, x -> {
                if (x.succeeded()) {
                    DomainEventStream stream = null;
                    Optional<JsonObject> first = x.result().getRows().stream().findFirst();
                    if (first.isPresent()) {
                        StreamRecord record = first.get().mapTo(StreamRecord.class);
                        stream = convertFrom(record);
                    }
                    future.complete(stream);
                    return;
                }
                future.completeExceptionally(x.cause());
            });
            return future.exceptionally(throwable -> {
                if (throwable instanceof SQLException) {
                    SQLException ex = (SQLException) throwable;
                    logger.error("Find event by commandId has sql exception, aggregateRootId: {}, commandId: {}", aggregateRootId, commandId, ex);
                    throw new IORuntimeException(throwable);
                }
                logger.error("Find event by commandId has unknown exception, aggregateRootId: {}, commandId: {}", aggregateRootId, commandId, throwable);
                throw new EventStoreException(throwable);
            });
        }, "FindEventByCommandIdAsync");

    }

    @Override
    public CompletableFuture<Integer> getPublishedVersionAsync(String aggregateRootTypeName, String aggregateRootId) {
        return IOHelper.tryIOFuncAsync(() -> {
            CompletableFuture<Integer> future = new CompletableFuture<>();
            String sql = String.format(SELECT_MAX_VERSION_SQL, getTableName(aggregateRootId));
            JsonArray array = new JsonArray();
            array.add(aggregateRootId);
            array.add(aggregateRootTypeName);
            sqlClient.querySingleWithParams(sql, array, x -> {
                if (x.succeeded()) {
                    int result = 0;
                    if (x.result() != null && x.result().size() > 0) {
                        result = Optional.ofNullable(x.result().getInteger(0)).orElse(0);
                    }
                    future.complete(result);
                    return;
                }
                future.completeExceptionally(x.cause());
            });
            return future.exceptionally(throwable -> {
                if (throwable instanceof SQLException) {
                    SQLException ex = (SQLException) throwable;
                    logger.error("Find version by aggregateRootId has sql exception, aggregateRootId: {}, aggregateRootTypeName: {}", aggregateRootId, aggregateRootTypeName, ex);
                    throw new IORuntimeException(throwable);
                }
                logger.error("Find version by aggregateRootId has unknown exception, aggregateRootId: {}, aggregateRootTypeName: {}", aggregateRootId, aggregateRootTypeName, throwable);
                throw new EventStoreException(throwable);
            });
        }, "getPublishedVersionAsync");
    }


    private int getTableIndex(String aggregateRootId) {
        int hash = aggregateRootId.hashCode();
        if (hash < 0) {
            hash = Math.abs(hash);
        }
        return hash % tableCount;
    }

    private String getTableName(String aggregateRootId) {
        if (tableCount <= 1) {
            return tableName;
        }
        int tableIndex = getTableIndex(aggregateRootId);
        return String.format(EVENT_TABLE_NAME_FORMAT, tableName, tableIndex);
    }

    private DomainEventStream convertFrom(StreamRecord record) {
        return new DomainEventStream(
                record.commandId,
                record.aggregateRootId,
                record.aggregateRootTypeName,
                record.gmtCreated,
                eventSerializer.deserialize(serializeService.deserialize(record.events, Map.class)),
                Maps.newHashMap());
    }

    private SQLClient batchWithParams(String sql, List<JsonArray> params, Handler<AsyncResult<List<Integer>>> handler) {
        sqlClient.getConnection(getConnection -> {
            if (getConnection.failed()) {
                handler.handle(Future.failedFuture(getConnection.cause()));
                return;
            }
            final SQLConnection conn = getConnection.result();
            conn.setAutoCommit(false, autocommit -> {
                if (autocommit.failed()) {
                    handler.handle(Future.failedFuture(autocommit.cause()));
                    resetAutoCommitAndCloseConnection(conn);
                    return;
                }
                conn.batchWithParams(sql, params, batch -> {
                    if (batch.succeeded()) {
                        conn.commit(commit -> {
                            if (commit.succeeded()) {
                                handler.handle(Future.succeededFuture(batch.result()));
                            } else {
                                handler.handle(Future.failedFuture(commit.cause()));
                            }
                            resetAutoCommitAndCloseConnection(conn);
                        });
                    } else {
                        conn.rollback(rollback -> {
                            if (rollback.succeeded()) {
                                handler.handle(Future.failedFuture(batch.cause()));
                            } else {
                                handler.handle(Future.failedFuture(rollback.cause()));
                            }
                            resetAutoCommitAndCloseConnection(conn);
                        });
                    }
                });
            });
        });
        return sqlClient;
    }

    private void resetAutoCommitAndCloseConnection(SQLConnection conn) {
        conn.setAutoCommit(true, commit -> {
            if (commit.failed()) {
                logger.error("jdbc driver set autocommit true failed", commit.cause());
            }
            // close will put the connection into pool
            conn.close();
        });
    }
}

File: jdbc/src/main/java/org/enodeframework/jdbc/JDBCEventStore.kt
package org.enodeframework.jdbc

import com.google.common.base.Strings
import com.google.common.collect.Lists
import com.google.common.collect.Maps
import io.vertx.core.AbstractVerticle
import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.ext.jdbc.JDBCClient
import io.vertx.ext.sql.ResultSet
import io.vertx.ext.sql.SQLClient
import io.vertx.ext.sql.SQLConnection
import org.enodeframework.configurations.DbType
import org.enodeframework.configurations.EventStoreConfiguration
import org.enodeframework.common.exception.EventStoreException
import org.enodeframework.common.exception.IORuntimeException
import org.enodeframework.common.io.IOHelper
import org.enodeframework.common.serializing.ISerializeService
import org.enodeframework.eventing.*
import org.slf4j.LoggerFactory
import java.sql.SQLException
import java.time.ZonedDateTime
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.regex.Pattern
import java.util.stream.Collectors
import javax.sql.DataSource

/**
 * @author anruence@gmail.com
 */
class JDBCEventStore(
    private val dataSource: DataSource,
    private val configuration: EventStoreConfiguration,
    private val eventSerializer: IEventSerializer,
    private val serializeService: ISerializeService
) : AbstractVerticle(), IEventStore {

    private lateinit var sqlClient: SQLClient

    private var code: String = ""

    override fun start() {
        super.start()
        this.sqlClient = JDBCClient.create(vertx, this.dataSource)
    }

    override fun stop() {
        super.stop()
        this.sqlClient.close()
    }

    override fun batchAppendAsync(eventStreams: List<DomainEventStream>): CompletableFuture<EventAppendResult> {
        val future = CompletableFuture<EventAppendResult>()
        val appendResult = EventAppendResult()
        if (eventStreams.isEmpty()) {
            future.complete(appendResult)
            return future
        }
        val eventStreamMap = eventStreams.stream().distinct()
            .collect(Collectors.groupingBy { obj: DomainEventStream -> obj.aggregateRootId })
        val batchAggregateEventAppendResult = BatchAggregateEventAppendResult(eventStreamMap.keys.size)
        for ((key, value) in eventStreamMap) {
            batchAppendAggregateEventsAsync(key, value, batchAggregateEventAppendResult, 0)
        }
        return batchAggregateEventAppendResult.taskCompletionSource
    }

    private fun batchAppendAggregateEventsAsync(
        aggregateRootId: String,
        eventStreamList: List<DomainEventStream>,
        batchAggregateEventAppendResult: BatchAggregateEventAppendResult,
        retryTimes: Int
    ) {
        IOHelper.tryAsyncActionRecursively(
            "BatchAppendAggregateEventsAsync",
            { batchAppendAggregateEventsAsync(aggregateRootId, eventStreamList) },
            { result: AggregateEventAppendResult? ->
                batchAggregateEventAppendResult.addCompleteAggregate(
                    aggregateRootId,
                    result
                )
            },
            { String.format("[aggregateRootId: %s, eventStreamCount: %s]", aggregateRootId, eventStreamList.size) },
            null,
            retryTimes, true
        )
    }

    private fun tryFindEventByCommandIdAsync(
        aggregateRootId: String,
        commandId: String,
        duplicateCommandIds: MutableList<String>,
        retryTimes: Int
    ): CompletableFuture<DomainEventStream?> {
        val future = CompletableFuture<DomainEventStream?>()
        IOHelper.tryAsyncActionRecursively(
            "TryFindEventByCommandIdAsync",
            { findAsync(aggregateRootId, commandId) },
            { result: DomainEventStream? ->
                if (result != null) {
                    duplicateCommandIds.add(result.commandId)
                }
                future.complete(result)
            },
            { String.format("[aggregateRootId:%s, commandId:%s]", aggregateRootId, commandId) },
            null,
            retryTimes, true
        )
        return future
    }

    private fun batchAppendAggregateEventsAsync(
        aggregateRootId: String,
        eventStreamList: List<DomainEventStream>
    ): CompletableFuture<AggregateEventAppendResult> {
        val sql = String.format(INSERT_EVENT_SQL, configuration.eventTableName)
        val jsonArrays: ArrayList<JsonArray> = Lists.newArrayList()
        for (domainEventStream in eventStreamList) {
            val array = JsonArray()
            array.add(domainEventStream.aggregateRootId)
            array.add(domainEventStream.aggregateRootTypeName)
            array.add(domainEventStream.commandId)
            array.add(domainEventStream.version)
            array.add(domainEventStream.timestamp.toInstant())
            array.add(serializeService.serialize(eventSerializer.serialize(domainEventStream.events())))
            jsonArrays.add(array)
        }
        val future = CompletableFuture<AggregateEventAppendResult>()
        batchWithParams(sql, jsonArrays) { ar: AsyncResult<List<Int?>?> ->
            if (ar.succeeded()) {
                val appendResult = AggregateEventAppendResult()
                appendResult.eventAppendStatus = EventAppendStatus.Success
                future.complete(appendResult)
                return@batchWithParams
            }
            val throwable = ar.cause()
            if (throwable is SQLException) {
                if (code == throwable.sqlState && throwable.message?.contains(configuration.eventVersionUkName) == true) {
                    val appendResult = AggregateEventAppendResult()
                    appendResult.eventAppendStatus = EventAppendStatus.DuplicateEvent
                    future.complete(appendResult)
                    return@batchWithParams
                }
                if (code == throwable.sqlState && throwable.message?.contains(configuration.eventCommandIdUkName) == true) {
                    // 不同的数据库在冲突时的错误信息不同，可以通过解析错误信息的方式将冲突的commandId找出来，这里要求id不能命中正则的规则（不包含-字符）
                    val appendResult = AggregateEventAppendResult()
                    appendResult.eventAppendStatus = EventAppendStatus.DuplicateCommand
                    val commandId = this.getDuplicatedId(throwable.message ?: "")
                    if (!Strings.isNullOrEmpty(commandId)) {
                        appendResult.duplicateCommandIds = Lists.newArrayList(commandId)
                    }
                    // 如果没有从异常信息获取到commandId(很低概率获取不到)，需要从db查询出来
                    // 但是Vert.x的线程模型决定了不能再次使用EventLoop线程执行阻塞的查询操作
                    future.complete(appendResult)
                    return@batchWithParams
                }
                logger.error("Batch append event has sql exception.", throwable)
                future.completeExceptionally(IORuntimeException(throwable))
                return@batchWithParams
            }
            logger.error("Batch append event has unknown exception.", throwable)
            future.completeExceptionally(EventStoreException(throwable))
            return@batchWithParams
        }
        return future
    }

    private fun tryFindEventByCommandIdAsyncRecursion(
        i: Int,
        aggregateRootId: String,
        eventStreamList: List<DomainEventStream>,
        duplicateCommandIds: MutableList<String>,
        retryTimes: Int
    ) {
        if (i == 0) {
            return
        }
        val eventStream = eventStreamList[i - 1]
        tryFindEventByCommandIdAsync(
            aggregateRootId,
            eventStream.commandId,
            duplicateCommandIds,
            retryTimes
        ).thenAccept { x: DomainEventStream? ->
            if (x != null) {
                return@thenAccept
            }
            tryFindEventByCommandIdAsyncRecursion(i - 1, aggregateRootId, eventStreamList, duplicateCommandIds, 0)
        }
    }

    override fun queryAggregateEventsAsync(
        aggregateRootId: String,
        aggregateRootTypeName: String,
        minVersion: Int,
        maxVersion: Int
    ): CompletableFuture<List<DomainEventStream>> {
        return IOHelper.tryIOFuncAsync({
            queryAggregateEvents(aggregateRootId, aggregateRootTypeName, minVersion, maxVersion)
        }, "QueryAggregateEventsAsync")
    }

    private fun queryAggregateEvents(
        aggregateRootId: String,
        aggregateRootTypeName: String,
        minVersion: Int,
        maxVersion: Int
    ): CompletableFuture<List<DomainEventStream>> {
        val future = CompletableFuture<List<DomainEventStream>>()
        val sql = String.format(SELECT_MANY_BY_VERSION_SQL, configuration.eventTableName)
        val array = JsonArray()
        array.add(aggregateRootId)
        array.add(minVersion)
        array.add(maxVersion)
        sqlClient.queryWithParams(sql, array) { ar: AsyncResult<ResultSet> ->
            if (ar.succeeded()) {
                future.complete(ar.result().rows.map { row ->
                    this.convertFrom(row)
                }.toMutableList())
                return@queryWithParams
            }
            val throwable = ar.cause()
            if (throwable is SQLException) {
                val errorMessage = String.format(
                    "Failed to query aggregate events async, aggregateRootId: %s, aggregateRootType: %s",
                    aggregateRootId,
                    aggregateRootTypeName
                )
                logger.error(errorMessage, throwable)
                future.completeExceptionally(IORuntimeException(throwable))
                return@queryWithParams
            }
            logger.error(
                "Failed to query aggregate events async, aggregateRootId: {}, aggregateRootType: {}",
                aggregateRootId,
                aggregateRootTypeName,
                throwable
            )
            future.completeExceptionally(EventStoreException(throwable))
            return@queryWithParams
        }
        return future
    }

    override fun findAsync(aggregateRootId: String, version: Int): CompletableFuture<DomainEventStream?> {
        return IOHelper.tryIOFuncAsync({
            findByVersion(aggregateRootId, version)
        }, "FindEventByVersionAsync")
    }

    private fun findByVersion(aggregateRootId: String, version: Int): CompletableFuture<DomainEventStream?> {
        val future = CompletableFuture<DomainEventStream?>()
        val sql = String.format(SELECT_ONE_BY_VERSION_SQL, configuration.eventTableName)
        val array = JsonArray()
        array.add(aggregateRootId)
        array.add(version)
        sqlClient.queryWithParams(sql, array) { ar: AsyncResult<ResultSet> ->
            if (ar.succeeded()) {
                ar.result().rows.firstOrNull().let { row ->
                    future.complete(row?.let { this.convertFrom(it) })
                    return@queryWithParams
                }
            }
            val throwable = ar.cause()
            if (throwable is SQLException) {
                logger.error(
                    "Find event by version has sql exception, aggregateRootId: {}, version: {}",
                    aggregateRootId,
                    version,
                    throwable
                )
                future.completeExceptionally(IORuntimeException(throwable))
                return@queryWithParams
            }
            logger.error(
                "Find event by version has unknown exception, aggregateRootId: {}, version: {}",
                aggregateRootId,
                version,
                throwable
            )
            future.completeExceptionally(EventStoreException(throwable))
            return@queryWithParams
        }
        return future
    }

    override fun findAsync(aggregateRootId: String, commandId: String): CompletableFuture<DomainEventStream?> {
        return IOHelper.tryIOFuncAsync({
            findByCommandId(aggregateRootId, commandId)
        }, "FindEventByCommandIdAsync")
    }

    private fun findByCommandId(aggregateRootId: String, commandId: String): CompletableFuture<DomainEventStream?> {
        val future = CompletableFuture<DomainEventStream?>()
        val sql = String.format(SELECT_ONE_BY_COMMAND_ID_SQL, configuration.eventTableName)
        val array = JsonArray()
        array.add(aggregateRootId)
        array.add(commandId)
        sqlClient.queryWithParams(sql, array) { ar: AsyncResult<ResultSet> ->
            if (ar.succeeded()) {
                ar.result().rows.firstOrNull().let { row ->
                    future.complete(row?.let { this.convertFrom(it) })
                    return@queryWithParams
                }
            }
            val throwable = ar.cause()
            if (throwable is SQLException) {
                logger.error(
                    "Find event by commandId has sql exception, aggregateRootId: {}, commandId: {}",
                    aggregateRootId,
                    commandId,
                    throwable
                )
                future.completeExceptionally(IORuntimeException(throwable))
                return@queryWithParams
            }
            logger.error(
                "Find event by commandId has unknown exception, aggregateRootId: {}, commandId: {}",
                aggregateRootId,
                commandId,
                throwable
            )
            future.completeExceptionally(EventStoreException(throwable))
            return@queryWithParams
        }
        return future

    }

    private fun convertFrom(record: JsonObject): DomainEventStream {
        return DomainEventStream(
            record.getString("command_id"),
            record.getString("aggregate_root_id"),
            record.getString("aggregate_root_type_name"),
            Date.from(ZonedDateTime.parse(record.getString("gmt_create")).toInstant()),
            eventSerializer.deserialize(
                serializeService.deserialize(
                    record.getString("events"),
                    MutableMap::class.java
                ) as MutableMap<String, String>
            ),
            Maps.newHashMap()
        )
    }

    private fun batchWithParams(
        sql: String,
        params: List<JsonArray>,
        handler: Handler<AsyncResult<List<Int?>?>>
    ): SQLClient {
        sqlClient.getConnection { getConnection: AsyncResult<SQLConnection> ->
            if (getConnection.failed()) {
                handler.handle(Future.failedFuture(getConnection.cause()))
                return@getConnection
            }
            val conn = getConnection.result()
            conn.setAutoCommit(false) { autocommit: AsyncResult<Void?> ->
                if (autocommit.failed()) {
                    handler.handle(Future.failedFuture(autocommit.cause()))
                    resetAutoCommitAndCloseConnection(conn)
                    return@setAutoCommit
                }
                conn.batchWithParams(sql, params) { batch: AsyncResult<List<Int?>?> ->
                    if (batch.succeeded()) {
                        conn.commit { commit: AsyncResult<Void?> ->
                            if (commit.succeeded()) {
                                handler.handle(Future.succeededFuture(batch.result()))
                            } else {
                                handler.handle(Future.failedFuture(commit.cause()))
                            }
                            resetAutoCommitAndCloseConnection(conn)
                        }
                    } else {
                        conn.rollback { rollback: AsyncResult<Void?> ->
                            if (rollback.succeeded()) {
                                handler.handle(Future.failedFuture(batch.cause()))
                            } else {
                                handler.handle(Future.failedFuture(rollback.cause()))
                            }
                            resetAutoCommitAndCloseConnection(conn)
                        }
                    }
                }
            }
        }
        return sqlClient
    }

    private fun resetAutoCommitAndCloseConnection(conn: SQLConnection) {
        conn.setAutoCommit(true) { commit: AsyncResult<Void?> ->
            if (commit.failed()) {
                logger.error("jdbc driver set autocommit true failed", commit.cause())
            }
            // close will put the connection into pool
            conn.close()
        }
    }

    // Duplicate entry '5d3ac841d1fcfe669e9a257d-5d3ac841d1fcfe669e9a2585' for key 'IX_EventStream_AggId_CommandId'
    // ERROR: duplicate key value violates unique constraint "event_stream_aggregate_root_id_command_id_key"\n详细：Key (aggregate_root_id, command_id)=(5ee99656d767113d73a7540f, 5ee99656d767113d73a75417) already exists.
    private fun getDuplicatedId(message: String): String {
        if (DbType.MySQL.name == configuration.dbType) {
            return getDuplicatedIdFromMySQL(message)
        }
        if (DbType.Pg.name == configuration.dbType) {
            return getDuplicatedIdFromPg(message)
        }
        return ""
    }

    private fun getDuplicatedIdFromMySQL(message: String): String {
        val matcher = PATTERN_MYSQL.matcher(message)
        if (!matcher.find()) {
            return ""
        }
        return if (matcher.groupCount() == 0) {
            ""
        } else matcher.group(1)
    }

    private fun getDuplicatedIdFromPg(message: String): String {
        val matcher = PATTERN_POSTGRESQL.matcher(message)
        if (!matcher.find()) {
            return ""
        }
        return if (matcher.groupCount() == 0) {
            ""
        } else matcher.group(1)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(JDBCEventStore::class.java)
        private const val INSERT_EVENT_SQL =
            "INSERT INTO %s (aggregate_root_id, aggregate_root_type_name, command_id, version, gmt_create, events) VALUES (?, ?, ?, ?, ?, ?)"
        private const val SELECT_MANY_BY_VERSION_SQL =
            "SELECT * FROM %s WHERE aggregate_root_id = ? AND version >= ? AND Version <= ? ORDER BY version"
        private const val SELECT_ONE_BY_VERSION_SQL = "SELECT * FROM %s WHERE aggregate_root_id = ? AND version = ?"
        private const val SELECT_ONE_BY_COMMAND_ID_SQL =
            "SELECT * FROM %s WHERE aggregate_root_id = ? AND command_id = ?"

        private val PATTERN_POSTGRESQL = Pattern.compile("=\\(.*, (.*)\\) already exists.$")
        private val PATTERN_MYSQL = Pattern.compile("^Duplicate entry '.*-(.*)' for key")
    }

    init {
        if (DbType.MySQL.name == configuration.dbType) {
            this.code = "23000"
        }
        if (DbType.Pg.name == configuration.dbType) {
            this.code = "23505"
        }
    }
}